### PR TITLE
hdf5@1.8: uses m4 for build

### PR DESCRIPTION
**Pull request by iMichka**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/hdf5@1.8.rb [hunk 1]
```diff
@@ -26,6 +26,8 @@ class Hdf5AT18 < Formula
   depends_on "gcc" # for gfortran
   depends_on "szip"
 
+  uses_from_macos "m4"
```
Comment (carlocab) [inline Formula/hdf5@1.8.rb:29]:
```suggestion
  uses_from_macos "m4" => :build
```
Given your title?

Comment (carlocab) [reply Formula/hdf5@1.8.rb:29]: Though given #73797 this may as well be a `depends_on`, possibly.

Comment (iMichka) [reply Formula/hdf5@1.8.rb]: Yes, you are right, I forgot to copy-paste it. Should be fixed now.

**Comment by iMichka** [comment]
@carlocab regarding #73797, I would recommend to make the change in that PR and modify all the uses_from_macos "m4" to depends_on. Or do you want to avoid having to rebase it in case this one is merged first?

**Comment by Bo98** [comment]
I'm not entirely sure why this is needed. `autoconf` already has a `uses_from_macos "m4"` so brewed `m4` should be installed.

**Comment by carlocab** [comment]
> modify all the uses_from_macos "m4" to depends_on.

Not quite all of them: just the ones with a (possibly indirect) dependency on `autoconf`.

**Comment by iMichka** [comment]
I'm running a test build on Linux here: https://github.com/Homebrew/linuxbrew-core/pull/22736

**Comment by iMichka** [comment]
```
2021-03-25T18:36:17.2541206Z [35m==>[0m [1mRunning Formulae#formula!(hdf5@1.8)[0m
2021-03-25T18:36:17.3152536Z [36m==>[0m [1mDetermining dependencies...[0m
2021-03-25T18:36:19.5549695Z [34m==>[0m [1mbrew fetch --retry autoconf automake berkeley-db binutils expat gcc gdbm gmp isl libmpc libtool m4 mpfr openssl@1.1 perl szip xz zlib[0m
2021-03-25T18:36:44.1294297Z [36m==>[0m [1mStarting build of hdf5@1.8[0m
2021-03-25T18:36:44.1299319Z [34m==>[0m [1mbrew fetch --retry hdf5@1.8 --build-bottle --force[0m
2021-03-25T18:36:49.5058379Z [34m==>[0m [1mbrew install --only-dependencies --verbose --build-bottle hdf5@1.8[0m
2021-03-25T18:37:59.1883541Z [34m==>[0m [1mbrew install --verbose --build-bottle hdf5@1.8[0m
2021-03-25T18:38:09.0205250Z [31m==>[0m [1m[31mFAILED[0m[0m
2021-03-25T18:38:09.0206965Z [34m==>[0m [1mDownloading https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.8/hdf5-1.8.22/src/hdf5-1.8.22.tar.bz2[0m
2021-03-25T18:38:09.0210207Z /usr/bin/curl --disable --globoff --show-error --user-agent Linuxbrew/3.0.9-20-g679dbea\ \(Linux\;\ x86_64\ Ubuntu\ 16.04.7\ LTS\)\ curl/7.47.0 --header Accept-Language:\ en --retry 3 --location --silent --head --request GET https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-1.8/hdf5-1.8.22/src/hdf5-1.8.22.tar.bz2
2021-03-25T18:38:09.0214877Z Already downloaded: /github/home/.cache/Homebrew/downloads/238c78407b5ce3d60979d2814cf24d6a76a0c7423c600ce4ce692d25915ec8d9--hdf5-1.8.22.tar.bz2
2021-03-25T18:38:09.0216780Z [34m==>[0m [1mVerifying checksum for '238c78407b5ce3d60979d2814cf24d6a76a0c7423c600ce4ce692d25915ec8d9--hdf5-1.8.22.tar.bz2'[0m
2021-03-25T18:38:09.0218881Z tar xof /github/home/.cache/Homebrew/downloads/238c78407b5ce3d60979d2814cf24d6a76a0c7423c600ce4ce692d25915ec8d9--hdf5-1.8.22.tar.bz2 -C /tmp/d20210325-5708-1hoyjl8
2021-03-25T18:38:09.0220548Z cp -pR /tmp/d20210325-5708-1hoyjl8/hdf5-1.8.22/. /tmp/hdf5-1.8-20210325-5708-l7f8gk/hdf5-1.8.22
2021-03-25T18:38:09.0221302Z chmod -Rf +w /tmp/d20210325-5708-1hoyjl8
2021-03-25T18:38:09.0221975Z [34m==>[0m [1mautoreconf -fiv[0m
2021-03-25T18:38:09.0222506Z autoreconf: Entering directory `.'
2021-03-25T18:38:09.0222988Z autoreconf: configure.ac: not using Gettext
2021-03-25T18:38:09.0223598Z autoreconf: running: aclocal --force 
2021-03-25T18:38:09.0224045Z autoreconf: configure.ac: tracing
2021-03-25T18:38:09.0224647Z autoreconf: running: libtoolize --copy --force
2021-03-25T18:38:09.0225160Z libtoolize:   error: One of these is required:
2021-03-25T18:38:09.0225594Z libtoolize:                 gm4 gnum4 m4
2021-03-25T18:38:09.0226550Z libtoolize:   error: Please install GNU M4, or 'export M4=/path/to/gnu/m4'.
2021-03-25T18:38:09.0227298Z autoreconf: libtoolize failed with exit status: 1
2021-03-25T18:38:09.0228294Z /usr/bin/ruby -e puts\ RUBY_VERSION
2021-03-25T18:38:09.0228928Z 
2021-03-25T18:38:09.0229378Z [34m==>[0m [1mFormula[0m
2021-03-25T18:38:09.0231063Z Path: /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/hdf5@1.8.rb
```

**Comment by Bo98** [comment]
Thanks, I'll see if I'll have a look and see what's happening.

**Comment by carlocab** [comment]
Ah. This was broken by https://github.com/Homebrew/brew/commit/f6eba3c1ea2c5ac069b667d45e39b739366e2ed6. I knew it would not do any harm 😄

**Comment by Bo98** [comment]
`m4` should be found on the path anyway.

**Comment by carlocab** [comment]
Does superenv also include recursive dependencies in `PATH`? I thought it only included direct ones.

**Comment by Bo98** [comment]
Ah yes, that'll be why. Forgot about that.

**Comment by carlocab** [comment]
So I guess we can go back to setting the `M4` environment variable, or just adding a dependency on `m4` whenever something breaks because it couldn't be found.

**Comment by Bo98** [comment]
> just adding a dependency on `m4` whenever something breaks because it couldn't be found.

This'll be anything that uses `libtool`. `autoconf` itself can cope because it has a baked in fallback path.

> Or defining `M4` manually in formulae.

I'd rather not do this. Either we go the route of depending on `m4` everywhere we use `libtool` or we provide it in the superenv.

**Comment by carlocab** [comment]
superenv seems like the cleanest solution. I'm not sure there's much of a point repeating code all over the place when it can be handled by superenv.

**Comment by carlocab** [comment]
Do we have any formulae that depend directly on `libtool` but not also on `autoconf`? (I obviously don't really understand autotools)

**Comment by Bo98** [comment]
Yes:

```
aubio
beecrypt
bochs
cfssl
corkscrew
dict
envoy
fontforge
gforth
graphicsmagick
guile@2
imagemagick
imagemagick@6
libextractor
libgetdata
libmp3splt
libprelude
libtermkey
libtommath
libvterm
libxo
mailutils
mesa-glu
nest
qemu
recode
sha1dc
srecord
stubby
synfig
tomcat-native
trafshow
unibilium
unixodbc
usbredir
yaf
yubico-piv-tool
```

**Comment by carlocab** [comment]
@Bo98

So what's the correct behaviour here? It seems like what you're suggesting is that we should be setting `M4` when there's a `libtool` dependency, rather than just an `autoconf` dependency.

**Comment by Bo98** [comment]
> It seems like what you're suggesting is that we should be setting `M4` when there's a `libtool` dependency, rather than just an `autoconf` dependency.

I'd say so, yes.

> we would probably need to have `libtool` depend on brewed `m4` as well, to avoid cases where `M4` points to macOS `m4` but the formula actually also depends on `autoconf`/`automake`.

I agree. Runtime dependency too - not just a build one like it is now on Linux.

**Comment by carlocab** [comment]
Actually, `libtool` declaring a runtime dependency on `m4` should be enough. There's no need to do anything in superenv. I just checked, and `m4` does appear in `PATH` at build time when a formula depends on `libtool` (and not also `m4`).

It didn't show up previously because it was only a build dependency.

I pushed the changed to #73797, so this PR will no longer be needed when that is merged.

**Comment by Bo98** [comment]
We don't add _any_ recursive dependencies of build dependencies currently.

**Comment by carlocab** [comment]
Yes, that's why it should suffice to make `m4` a runtime dep of `libtool`, no?

**Comment by Bo98** [comment]
No because even runtime deps of buildtime deps are not in the PATH.

**Comment by carlocab** [comment]
Gotcha. https://github.com/Homebrew/brew/pull/10944

**Pull request closed without merging**